File: Interview/flatten-2d-vector.java
//  Flatten 2D Vector

// Time O(N), Space O(1)
public class Vector2D implements Iterator<Integer> {
    Iterator<List<Integer>> listIter;
    Iterator<Integer> iter;
    
    public Vector2D(List<List<Integer>> vec2d) {
       listIter = vec2d.iterator();
    }

    @Override
    public Integer next() {
        return iter.next();
    }

    @Override
    public boolean hasNext() {
        while((iter==null || !iter.hasNext()) && (listIter!=null && listIter.hasNext())){
            iter= listIter.next().iterator();
        }
        return (iter!=null && iter.hasNext());
    }
}
/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i = new Vector2D(vec2d);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: Design/flatten-2d-vector.cc
//  Flatten 2D Vector

class Vector2D {
public:
    Vector2D(vector<vector<int>>& vec2d) {
        vec_begin = vec2d.begin();
        vec_end = vec2d.end();
        row = vec_begin;
        if (row != vec_end)
            col = row->begin();
    }

    int next() {
        int res = *col;
        col ++;
        return res;
    }

    bool hasNext() {
        if (row == vec_end)  return false;
        if (col == row->end()) {
            row ++;
            for (; row!=vec_end && row->size()==0; row++);
            if (row == vec_end) return false;
            col = row->begin();
        }
        return true;
    }
    
private:
    vector<vector<int>>::iterator row;
    vector<int>::iterator col;
    vector<vector<int>>::iterator vec_begin, vec_end;
};

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D i(vec2d);
 * while (i.hasNext()) cout << i.next();
 */
